streaming/base/shared.py

```diff
@@ -27,37 +33,50 @@ class SharedBarrier:
     Args:
         filelock_path (str): Path to lock file on local filesystem.
         shm_path (str): Shared memory object name in /dev/shm.
+        is_local_leader (bool): Is a local leader process or not
     """
 
-    def __init__(self, filelock_path: str, shm_path: str) -> None:
+    def __init__(self, filelock_path: str, shm_path: str, is_local_leader: bool) -> None:
+        self.is_local_leader = is_local_leader
         self.filelock_path = filelock_path
         self.shm_path = shm_path
 
-        # Create filelock.
-        dirname = os.path.dirname(filelock_path)
-        os.makedirs(dirname, exist_ok=True)
-        self.lock = FileLock(filelock_path)
-
         # Create three int32 fields in shared memory: num_enter, num_exit, flag.
         size = 3 * np.int32(0).nbytes
+
         try:
+            # Creates a new shared memory block
             self._shm = SharedMemory(shm_path, True, size)
         except FileExistsError:
             sleep(TICK)
+            # Attaches to an existing shared memory block
             self._shm = SharedMemory(shm_path, False, size)
+
+        # Create filelock.
+        self.dirname = os.path.dirname(filelock_path)
+        os.makedirs(self.dirname, exist_ok=True)
+        self.lock = FileLock(filelock_path)
+
         self._arr = np.ndarray(3, buffer=self._shm.buf, dtype=np.int32)
         self._arr[0] = 0
         self._arr[1] = -1
         self._arr[2] = True
 
     def __del__(self):
         """Destructor clears array that references shm."""
-        try:
-            del self._arr
+        if hasattr(self, '_shm') and self._shm is not None:
+            # Close each SharedMemory instance
             self._shm.close()
-            self._shm.unlink()
-        except:
-            pass
+            if self.is_local_leader:
+                # Call unlink only once to release the shared memory
+                self._shm.unlink()
+            else:
+                # Wait for local leader process to execute first
+                sleep(1)
```

Comment:
Is this sleep necessary/does it work? It seems plausible that a worker process is more than 1 second off of the leader.

Reply:
Looks pretty harmless in either case to me?

Reply:
my concern was more if it was actually supposed to prevent some failure case, I'm not sure how effective it would be.

Reply:
"DDP, oh god why, M-O-U-S-E"

Reply:
Merging to unblock others, let's improve this afterward

Reply:
Yes, the sleep is necessary. We have a local rank0 barrier method called `wait_for_local_leader()` which waits for leader rank to be in sync, hence, there won't be a case most of the time that the local rank 0 is off. Also, if it is off by some seconds also then it won't create any harm. The purpose of adding those rank0 barrier and sleep is to avoid unnecessary warning message as shown below from rank 0. We are releasing all the resources, it's just how python track the resources and throws a warning from other rank processes thinking that resources was not released.

```
/usr/lib/python3.10/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/88fcf1_next_epoch': [Errno 2] No such file or directory: '/88fcf1_next_epoch'
  warnings.warn('resource_tracker: %r: %s' % (name, e))
/usr/lib/python3.10/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/6f2a6f_shard_states': [Errno 2] No such file or directory: '/6f2a6f_shard_states'
  warnings.warn('resource_tracker: %r: %s' % (name, e))
/usr/lib/python3.10/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/6f2a6f_barrier': [Errno 2] No such file or directory: '/6f2a6f_barrier'
  warnings.warn('resource_tracker: %r: %s' % (name, e))
```